akd/src/tree_node.rs

```diff
@@ -49,6 +49,141 @@ impl NodeType {
+/// we can't guarantee that a "publish" operation is globally atomic at the storage layer,
+/// however we do assume record-level atomicity. This means that some records may be updated
+/// to "future" values, and therefore we might need to temporarily read their previous values.
+///
+/// The Directory publishes the AZKS last of all, once all other records are successfully written.
+/// This single record is where the "current" epoch is determined, so until it's publish a Directory
+/// instance (example: read-only, only generating proofs) will think current history is set at epoch - 1.
+///
+/// This structure holds the label along with the current value & epoch - 1
+#[derive(Debug, Eq, PartialEq, Clone)]
+#[cfg_attr(
+    feature = "serde_serialization",
+    derive(serde::Deserialize, serde::Serialize)
+)]
+#[cfg_attr(feature = "serde_serialization", serde(bound = ""))]
+pub struct TreeNodeWithPreviousValue {
+    /// The label of the node
+    pub label: NodeLabel,
+    /// The "latest" node, either future or current
+    pub latest_node: TreeNode,
+    /// The "previous" node, either current or past
+    pub previous_node: Option<TreeNode>,
+}
+
+impl Storable for TreeNodeWithPreviousValue {
+    type Key = NodeKey;
+
+    fn data_type() -> StorageType {
+        StorageType::TreeNode
+    }
+
+    fn get_id(&self) -> NodeKey {
+        NodeKey(self.label)
+    }
+
+    fn get_full_binary_key_id(key: &NodeKey) -> Vec<u8> {
+        let mut result = vec![StorageType::TreeNode as u8];
+        result.extend_from_slice(&key.0.len.to_le_bytes());
+        result.extend_from_slice(&key.0.val);
+        result
+    }
+
+    fn key_from_full_binary(bin: &[u8]) -> Result<NodeKey, String> {
+        if bin.len() < 37 {
+            return Err("Not enough bytes to form a proper key".to_string());
+        }
+
+        if bin[0] != StorageType::TreeNode as u8 {
+            return Err("Not a history tree node key".to_string());
+        }
+
+        let len_bytes: [u8; 4] = bin[1..=4].try_into().expect("Slice with incorrect length");
+        let val_bytes: [u8; 32] = bin[5..=36].try_into().expect("Slice with incorrect length");
+        let len = u32::from_le_bytes(len_bytes);
+
+        Ok(NodeKey(NodeLabel::new(val_bytes, len)))
+    }
+}
+
+impl TreeNodeWithPreviousValue {
+    pub(crate) fn from_tree_node(node: TreeNode) -> Self {
+        Self {
+            label: node.label,
+            latest_node: node,
+            previous_node: None,
+        }
+    }
+
+    pub(crate) async fn write_to_storage<S: Storage + Send + Sync>(
+        &self,
+        storage: &S,
+    ) -> Result<(), StorageError> {
+        storage.set(DbRecord::TreeNode(self.clone())).await
+    }
+
+    pub(crate) async fn get_from_storage<S: Storage + Send + Sync>(
+        storage: &S,
+        key: &NodeKey,
+        current_epoch: u64,
+    ) -> Result<TreeNode, StorageError> {
```

Comment:
Yeah so for the normal get, that does make sense, but for the batch get it'd require a 2nd iteration of the vector since we'd only check if which node to retrieve in the TreeNode impl. I'm going to rename the functions however to be more logical, they were bad copy-pasta names initially as I was trying to work out the best way to do this shift-left logix.

Reply:
This is what I'm going to modify it to: 
```rust
pub(crate) async fn get_appropriate_tree_node_from_storage<S: Storage + Send + Sync>(
      storage: &S,
      key: &NodeKey,
      target_epoch: u64,
  ) -> Result<TreeNode, StorageError> {
      ...
  }

  pub(crate) async fn batch_get_appropriate_tree_node_from_storage<S: Storage + Send + Sync>(
      storage: &S,
      keys: &[NodeKey],
      target_epoch: u64,
  ) -> Result<Vec<TreeNode>, StorageError> {
      ...
  }
```